### PR TITLE
Add debug mode

### PR DESCRIPTION
**Pull request by driftluo**
1. disable discovery publish unreachable ip
2. add test mode to discovery and identify protocol, allow publish any ip

### CODE REVIEW

File: protocols/identify/src/lib.rs [hunk 1]
```diff
@@ -137,11 +145,12 @@ impl<T: Callback> IdentifyProtocol<T> {
                 .misbehave(&info.peer_id, Misbehavior::TooManyAddresses(listens.len()))
         } else {
             trace!("received listen addresses: {:?}", listens);
+            let debug = self.debug;
             let reachable_addrs = listens
                 .into_iter()
                 .filter(|addr| {
                     multiaddr_to_socketaddr(addr)
-                        .map(|socket_addr| is_reachable(socket_addr.ip()))
+                        .map(|socket_addr| debug || is_reachable(socket_addr.ip()))
```
Comment (TheWaWaR) [inline protocols/identify/src/lib.rs]: Better wrapper this logic `debug || is_reachable(socket_addr.ip())` in `self.is_reachable()`.

File: protocols/discovery/src/lib.rs [hunk 2]
```diff
@@ -290,8 +300,12 @@ impl<M: AddressManager> Stream for Discovery<M> {
 
         let mut dead_addr = Vec::default();
         for key in self.dead_keys.drain() {
-            if let Some(value) = self.substreams.remove(&key) {
-                dead_addr.push(RawAddr::from(value.remote_addr.into_inner()))
+            if let Some(Some(addr)) = self
+                .substreams
+                .remove(&key)
+                .map(|value| value.remote_addr.into_inner())
```
Comment (TheWaWaR) [inline protocols/discovery/src/lib.rs]: Use `and_then`

File: protocols/identify/src/lib.rs (hunk 1 above)
Comment (driftluo) [reply protocols/identify/src/lib.rs]: Can't do this, borrow check gank you

Comment (TheWaWaR) [reply protocols/identify/src/lib.rs]: Fine, then.

File: protocols/discovery/src/substream.rs [hunk 3]
```diff
@@ -129,16 +129,33 @@ impl SubstreamValue {
...
         max_known: usize,
         query_cycle: Option<Duration>,
+        global_ip_only: bool,
     ) -> SubstreamValue {
         let session_id = substream.stream.session_id;
         let mut pending_messages = VecDeque::default();
         debug!("direction: {:?}", direction);
         let mut addr_known = AddrKnown::new(max_known);
-        let remote_addr = if direction.is_outbound() {
+        let is_reachable = !global_ip_only
```
Comment (quake) [inline protocols/discovery/src/substream.rs]: Can we keep the same code in this block and move the `global_ip_only` checking to line 242 ?

**Pull request merged**